### PR TITLE
fix(datafile-parsing): Prevent newer versions datafile

### PR DESCRIPTION
**Pull request by msohailhussain**

### CODE REVIEW

File: OptimizelySDK/Optimizely.cs [hunk 1]
```diff
@@ -111,9 +112,15 @@ public class Optimizely : IOptimizely
                 IsValid = true;
                 DecisionService = new DecisionService(Bucketer, ErrorHandler, Config, userProfileService, Logger);
             }
+            catch (ConfigParseException configException)
+            {
+                Logger.Log(LogLevel.ERROR, configException.Message);
```
[the remark this answers is not in the document]
Comment (mfahadahmed) [reply OptimizelySDK/Optimizely.cs]: Done.

[the remark this answers is not in the document]
Comment (mikeng13) [reply OptimizelySDK/Optimizely.cs]: Let's consolidate these two catches into one and handle the logging based on the type of the exception. We can avoid duplicating some of the code, like passing the error to the error handler

File: OptimizelySDK/ProjectConfig.cs [hunk 2]
```diff
@@ -280,6 +296,24 @@ public static ProjectConfig Create(string content, ILogger logger, IErrorHandler
...
+        {
+            if (configData == null)
+                throw new ConfigParseException("Unable to parse null datafile.");
+
+            if (string.IsNullOrEmpty(configData))
+                throw new ConfigParseException("Unable to parse empty datafile.");
+
+            var config = JsonConvert.DeserializeObject<ProjectConfig>(configData);
+
+            if (SupportedVersions.TrueForAll((obj) => !(((int)obj).ToString() == config.Version)))
```
Comment (mikeng13) [inline OptimizelySDK/ProjectConfig.cs]: Please don't use the generic variable name `obj`. Instead go for `supportVersion`

File: OptimizelySDK.Tests/ProjectConfigTest.cs [hunk 3]
```diff
@@ -863,5 +863,36 @@ public void TestGetAttributeIdWithInvalidAttributeKey()
...
+        public void TestCreateThrowsWithEmptyDatafile()
+        {
+            var exception = Assert.Throws<ConfigParseException>(() => ProjectConfig.Create("", null, null));
+            Assert.AreEqual("Unable to parse empty datafile.", exception.Message);
+        }
+
+        [Test]
+        public void TestCreateThrowsWithUnsupportedDatafileVersion()
+        {
+            var unsupportedConfig = JsonConvert.DeserializeObject<ProjectConfig>(TestData.Datafile);
```
Comment (mikeng13) [inline OptimizelySDK.Tests/ProjectConfigTest.cs]: Please create a separate mock unsupported datafile and load it into this test case

File: OptimizelySDK.Tests/OptimizelyTest.cs [hunk 4]
```diff
@@ -185,7 +186,23 @@ public void TestValidateInputsInvalidFileJsonValidationSkipped()
         {
             string datafile = "{\"name\":\"optimizely\"}";
             Optimizely optimizely = new Optimizely(datafile, null, null, null, skipJsonValidation: true);
-            Assert.IsTrue(optimizely.IsValid);
+            Assert.IsFalse(optimizely.IsValid);
+        }
+
+        [Test]
+        public void TestErrorHandlingWithInvalidConfigVersion()
```
Comment (mikeng13) [inline OptimizelySDK.Tests/OptimizelyTest.cs]: Please split these up into three separate unit tests. Makes it easier to read and determine what actually failed

**Pull request merged**